internal/definitions/definition_handler_identity_claim.go

```diff
@@ -0,0 +1,211 @@
+// Copyright © 2022 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package definitions
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hyperledger/firefly/internal/log"
+	"github.com/hyperledger/firefly/pkg/database"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+)
+
+func (dh *definitionHandlers) handleIdentityClaimBroadcast(ctx context.Context, state DefinitionBatchState, msg *fftypes.Message, data []*fftypes.Data, verificationID *fftypes.UUID) (DefinitionMessageAction, error) {
+	var claim fftypes.IdentityClaim
+	valid := dh.getSystemBroadcastPayload(ctx, msg, data, &claim)
+	if !valid {
+		return ActionReject, nil
+	}
+
+	return dh.handleIdentityClaim(ctx, state, msg, &claim, verificationID)
+
+}
+
+func (dh *definitionHandlers) verifyClaimSignature(ctx context.Context, msg *fftypes.Message, identity *fftypes.Identity, parent *fftypes.Identity) (valid bool) {
+
+	author := msg.Header.Author
+	if author == "" {
+		return false
+	}
+
+	var expectedSigner *fftypes.Identity
+	switch {
+	case identity.Type == fftypes.IdentityTypeNode:
+		// In the special case of a node, the parent signs it directly
+		expectedSigner = parent
+	default:
+		expectedSigner = identity
+	}
+
+	valid = author == expectedSigner.DID ||
+		(expectedSigner.Type == fftypes.IdentityTypeOrg || author == fmt.Sprintf("%s%s", fftypes.FireFlyOrgDIDPrefix, expectedSigner.ID))
+	if !valid {
+		log.L(ctx).Warnf("Unable to process identity claim %s - signature mismatch type=%s author=%s expected=%s", msg.Header.ID, identity.Type, author, expectedSigner.DID)
+	}
+	return valid
+}
+
+func (dh *definitionHandlers) getClaimVerifier(msg *fftypes.Message, identity *fftypes.Identity) *fftypes.Verifier {
+	verifier := &fftypes.Verifier{
+		ID:        fftypes.NewUUID(),
+		Identity:  identity.ID,
+		Namespace: identity.Namespace,
+	}
+	switch identity.Type {
+	case fftypes.IdentityTypeNode:
+		verifier.VerifierRef.Type = fftypes.VerifierTypeFFDXPeerID
+		verifier.VerifierRef.Value = identity.Profile.GetString("id")
+	default:
+		verifier.VerifierRef.Type = dh.blockchain.VerifierType()
+		verifier.VerifierRef.Value = msg.Header.Key
+	}
+	return verifier
+}
+
+func (dh *definitionHandlers) confirmVerificationForClaim(ctx context.Context, state DefinitionBatchState, msg *fftypes.Message, identity, parent *fftypes.Identity) (*fftypes.UUID, error) {
```

Comment:
Note this function has to work in two cases, because we cannot guarantee the batch-pin order of two transactions signed by different blockchain identities (they will be batched separately, and ordered into blocks separately):

1. The claim comes in first before the verification
    - In this case, the claim message is `confirmed` and this function will be run again when the `verification` comes in
2. The verification comes in first
    - The verification is `confirmed` when it arrives, so that we can search for it as confirmed here when we (later) process the verification

Note we handle the case where the claim and the verification end up in the same definition-batch, which required an update to the `DefinitionBatchState` interface (more later on that).